In [0]:
pip install pomegranate

     |████████████████████████████████| 3.3MB 4.8MB/s 
  Created wheel for pomegranate: filename=pomegranate-0.12.2-cp36-cp36m-linux_x86_64.whl size=10965195 sha256=5a9eda34fe81e1b85b11c1682c02f28a940f11574bffe3a20e62ca963d905851
  Stored in directory: /root/.cache/pip/wheels/51/14/41/7285a60a08951abde03ce60f47a8644dd96d8e22b373bbbcc8
Successfully built pomegranate


In [0]:
pip install ipython-autotime

  Created wheel for ipython-autotime: filename=ipython_autotime-0.1-cp36-none-any.whl size=1832 sha256=012ddfadc122d30cf1fcd3551e9991a08182b1b1f198efad03958edd163fa64b
  Stored in directory: /root/.cache/pip/wheels/d2/df/81/2db1e54bc91002cec40334629bc39cfa86dff540b304ebcd6e
Successfully built ipython-autotime


In [0]:
%load_ext autotime

In [0]:
import time
import numpy as np
import pandas as pd
from pomegranate import *
import matplotlib.pyplot as plt
import random

time: 4.05 s


In [0]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).
time: 1.91 ms


In [0]:
dataset = pd.read_csv("/content/drive/My Drive/DAMIN/drug_consumption - Copy.csv")
dataset

ID      Age   Gender  Education  ...  Mushrooms  Nicotine  Semer  VSA
0        1  0.49788  0.48246   -0.05921  ...        CL0       CL2    CL0  CL0
1        2 -0.07854 -0.48246    1.98437  ...        CL0       CL4    CL0  CL0
2        3  0.49788 -0.48246   -0.05921  ...        CL1       CL0    CL0  CL0
3        4 -0.95197  0.48246    1.16365  ...        CL0       CL2    CL0  CL0
4        5  0.49788  0.48246    1.98437  ...        CL2       CL2    CL0  CL0
...    ...      ...      ...        ...  ...        ...       ...    ...  ...
1880  1884 -0.95197  0.48246   -0.61113  ...        CL0       CL0    CL0  CL5
1881  1885 -0.95197 -0.48246   -0.61113  ...        CL4       CL5    CL0  CL0
1882  1886 -0.07854  0.48246    0.45468  ...        CL2       CL6    CL0  CL0
1883  1887 -0.95197  0.48246   -0.61113  ...        CL3       CL4    CL0  CL0
1884  1888 -0.95197 -0.48246   -0.61113  ...        CL3       CL6    CL0  CL2

[1885 rows x 32 columns]

time: 109 ms


In [0]:
#Hanya gunakan non categorical values
df = dataset[['Age', 'Nscore', 'Escore', 'Oscore', 'Ascore', 'Cscore', 'Impulsive', 'SS']]
df

Age   Nscore   Escore   Oscore   Ascore   Cscore  Impulsive       SS
0     0.49788  0.31287 -0.57545 -0.58331 -0.91699 -0.00665   -0.21712 -1.18084
1    -0.07854 -0.67825  1.93886  1.43533  0.76096 -0.14277   -0.71126 -0.21575
2     0.49788 -0.46725  0.80523 -0.84732 -1.62090 -1.01450   -1.37983  0.40148
3    -0.95197 -0.14882 -0.80615 -0.01928  0.59042  0.58489   -1.37983 -1.18084
4     0.49788  0.73545 -1.63340 -0.45174 -0.30172  1.30612   -0.21712 -0.21575
...       ...      ...      ...      ...      ...      ...        ...      ...
1880 -0.95197 -1.19430  1.74091  1.88511  0.76096 -1.13788    0.88113  1.92173
1881 -0.95197 -0.24649  1.74091  0.58331  0.76096 -1.51840    0.88113  0.76540
1882 -0.07854  1.13281 -1.37639 -1.27553 -1.77200 -1.38502    0.52975 -0.52593
1883 -0.95197  0.91093 -1.92173  0.29338 -1.62090 -2.57309    1.29221  1.22470
1884 -0.95197 -0.46725  2.12700  1.65653  1.11406  0.41594    0.88113  1.22470

[1885 rows x 8 columns]

time: 27 ms


In [0]:
X = df.values
df

Age   Nscore   Escore   Oscore   Ascore   Cscore  Impulsive       SS
0     0.49788  0.31287 -0.57545 -0.58331 -0.91699 -0.00665   -0.21712 -1.18084
1    -0.07854 -0.67825  1.93886  1.43533  0.76096 -0.14277   -0.71126 -0.21575
2     0.49788 -0.46725  0.80523 -0.84732 -1.62090 -1.01450   -1.37983  0.40148
3    -0.95197 -0.14882 -0.80615 -0.01928  0.59042  0.58489   -1.37983 -1.18084
4     0.49788  0.73545 -1.63340 -0.45174 -0.30172  1.30612   -0.21712 -0.21575
...       ...      ...      ...      ...      ...      ...        ...      ...
1880 -0.95197 -1.19430  1.74091  1.88511  0.76096 -1.13788    0.88113  1.92173
1881 -0.95197 -0.24649  1.74091  0.58331  0.76096 -1.51840    0.88113  0.76540
1882 -0.07854  1.13281 -1.37639 -1.27553 -1.77200 -1.38502    0.52975 -0.52593
1883 -0.95197  0.91093 -1.92173  0.29338 -1.62090 -2.57309    1.29221  1.22470
1884 -0.95197 -0.46725  2.12700  1.65653  1.11406  0.41594    0.88113  1.22470

[1885 rows x 8 columns]

time: 29.4 ms


In [0]:
X.shape

(1885, 8)

time: 2.31 ms


#KNN

In [0]:
#membuat NaN values ke dataset
random.seed(0)
np.random.seed(0)
np.set_printoptions(suppress=True)

idxs = numpy.random.choice(1885, replace=False, size=50)
j =  np.random.choice(range(4,8),50)
X_nan = X.copy()
print(idxs)
print(j)

[ 220  506 1758 1678   76   27  773 1223  384 1769   85 1476 1222 1520
  539  289 1533  708  459 1444 1771  574  107 1727  521  578   18 1110
  582 1258  744 1497 1364   80 1055  440 1642 1635 1232 1881 1440 1093
  745 1449 1814  776 1547  473 1051   53]
[4 5 7 7 7 5 6 7 5 4 4 7 6 4 5 5 4 7 7 4 5 6 5 6 6 4 7 6 5 7 7 5 4 7 6 5 6
 5 5 7 5 7 5 7 7 4 7 7 7 4]
time: 4.79 ms


In [0]:
X_nan[idxs, j] = np.nan
X_nan[220][4]

nan

time: 3.38 ms


In [0]:
from sklearn.impute import KNNImputer

imputer = KNNImputer(n_neighbors=20)
df_filled = imputer.fit_transform(X_nan)
df_filled

array([[ 0.49788,  0.31287, -0.57545, ..., -0.00665, -0.21712, -1.18084],
       [-0.07854, -0.67825,  1.93886, ..., -0.14277, -0.71126, -0.21575],
       [ 0.49788, -0.46725,  0.80523, ..., -1.0145 , -1.37983,  0.40148],
       ...,
       [-0.07854,  1.13281, -1.37639, ..., -1.38502,  0.52975, -0.52593],
       [-0.95197,  0.91093, -1.92173, ..., -2.57309,  1.29221,  1.2247 ],
       [-0.95197, -0.46725,  2.127  , ...,  0.41594,  0.88113,  1.2247 ]])

time: 106 ms


In [0]:
#check error
import sklearn.metrics as metrics

mae = metrics.mean_absolute_error(X, df_filled)
mse = metrics.mean_squared_error(X, df_filled)
rmse = np.sqrt(mse) # or mse**(0.5)

print("Results of sklearn.metrics:")
print("MAE:",mae)
print("MSE:", mse)
print("RMSE:", rmse)

Results of sklearn.metrics:
MAE: 0.0016458772877984084
MSE: 0.0014636197483123177
RMSE: 0.03825728359818974
time: 6.86 ms


#Bayesian

In [0]:
#membuat NaN values ke dataset
random.seed(0)
np.random.seed(0)
np.set_printoptions(suppress=True)

idxs = numpy.random.choice(1885, replace=False, size=50)
j =  np.random.choice(range(4,8),50)
X_nan = X.copy()
print(idxs)
print(j)
X_nan[idxs, j] = numpy.nan

[ 220  506 1758 1678   76   27  773 1223  384 1769   85 1476 1222 1520
  539  289 1533  708  459 1444 1771  574  107 1727  521  578   18 1110
  582 1258  744 1497 1364   80 1055  440 1642 1635 1232 1881 1440 1093
  745 1449 1814  776 1547  473 1051   53]
[4 5 7 7 7 5 6 7 5 4 4 7 6 4 5 5 4 7 7 4 5 6 5 6 6 4 7 6 5 7 7 5 4 7 6 5 6
 5 5 7 5 7 5 7 7 4 7 7 7 4]
time: 6.02 ms


In [0]:
X_nan.shape

(1885, 8)

time: 2.35 ms


In [0]:
model1 = BayesianNetwork.from_samples(X_nan, max_parents=5)

time: 29.7 s


In [0]:
model1.predict_proba({})

array([{
    "class" :"Distribution",
    "dtype" :"numpy.float64",
    "name" :"DiscreteDistribution",
    "parameters" :[
        {
            "0.49788000000000004" :0.18885941644562335,
            "-0.07854" :0.25517241379310335,
            "-0.9519700000000001" :0.3411140583554375,
            "2.59171" :0.009549071618037346,
            "1.09449" :0.15596816976127323,
            "1.82213" :0.049336870026525356
        }
    ],
    "frozen" :false
},
       {
    "class" :"Distribution",
    "dtype" :"numpy.float64",
    "name" :"DiscreteDistribution",
    "parameters" :[
        {
            "0.31287" :0.035013262599469325,
            "-0.67825" :0.03183023872679032,
            "-0.46725" :0.04615384615384586,
            "-0.14882" :0.04031830238726768,
            "0.73545" :0.02599469496021213,
            "-1.32828" :0.018567639257294443,
            "0.62967" :0.040848806366047506,
            "-0.24649000000000001" :0.036074270557029,
            "-1.05308" :0.0302387

time: 16.4 ms


In [0]:
y_pred1 = model1.predict(X_nan)

time: 7.66 s


In [0]:
#check error
import sklearn.metrics as metrics

mae = metrics.mean_absolute_error(X, np.array(y_pred1))
mse = metrics.mean_squared_error(X, np.array(y_pred1))
rmse = np.sqrt(mse) # or mse**(0.5)

print("Results of sklearn.metrics:")
print("MAE:",mae)
print("MSE:", mse)
print("RMSE:", rmse)

Results of sklearn.metrics:
MAE: 0.002153342838196286
MSE: 0.0026644241845424407
RMSE: 0.05161806064298077
time: 7.96 ms


#Deterministic Regression

In [0]:
df2 = dataset[['Age', 'Nscore', 'Escore', 'Oscore', 'Ascore', 'Cscore', 'Impulsive', 'SS']]
# df2 = pd.get_dummies(df2, columns=['Alcohol'])
X2 = df2.values
df2

Age   Nscore   Escore   Oscore   Ascore   Cscore  Impulsive       SS
0     0.49788  0.31287 -0.57545 -0.58331 -0.91699 -0.00665   -0.21712 -1.18084
1    -0.07854 -0.67825  1.93886  1.43533  0.76096 -0.14277   -0.71126 -0.21575
2     0.49788 -0.46725  0.80523 -0.84732 -1.62090 -1.01450   -1.37983  0.40148
3    -0.95197 -0.14882 -0.80615 -0.01928  0.59042  0.58489   -1.37983 -1.18084
4     0.49788  0.73545 -1.63340 -0.45174 -0.30172  1.30612   -0.21712 -0.21575
...       ...      ...      ...      ...      ...      ...        ...      ...
1880 -0.95197 -1.19430  1.74091  1.88511  0.76096 -1.13788    0.88113  1.92173
1881 -0.95197 -0.24649  1.74091  0.58331  0.76096 -1.51840    0.88113  0.76540
1882 -0.07854  1.13281 -1.37639 -1.27553 -1.77200 -1.38502    0.52975 -0.52593
1883 -0.95197  0.91093 -1.92173  0.29338 -1.62090 -2.57309    1.29221  1.22470
1884 -0.95197 -0.46725  2.12700  1.65653  1.11406  0.41594    0.88113  1.22470

[1885 rows x 8 columns]

time: 29.2 ms


In [0]:
#membuat NaN values ke dataset
random.seed(0)
np.random.seed(0)
np.set_printoptions(suppress=True)

idxs = np.random.choice(1885, replace=False, size=50)
j =  np.random.choice(range(4,8),50)
X_nan = X2.copy()
print(idxs)
print(j)
X_nan[idxs, j] = numpy.nan
df_nan = pd.DataFrame(X_nan, columns=['Age', 'Nscore', 'Escore', 'Oscore', 'Ascore', 'Cscore', 'Impulsive', 'SS'])
df_nan

[ 220  506 1758 1678   76   27  773 1223  384 1769   85 1476 1222 1520
  539  289 1533  708  459 1444 1771  574  107 1727  521  578   18 1110
  582 1258  744 1497 1364   80 1055  440 1642 1635 1232 1881 1440 1093
  745 1449 1814  776 1547  473 1051   53]
[4 5 7 7 7 5 6 7 5 4 4 7 6 4 5 5 4 7 7 4 5 6 5 6 6 4 7 6 5 7 7 5 4 7 6 5 6
 5 5 7 5 7 5 7 7 4 7 7 7 4]


Age   Nscore   Escore   Oscore   Ascore   Cscore  Impulsive       SS
0     0.49788  0.31287 -0.57545 -0.58331 -0.91699 -0.00665   -0.21712 -1.18084
1    -0.07854 -0.67825  1.93886  1.43533  0.76096 -0.14277   -0.71126 -0.21575
2     0.49788 -0.46725  0.80523 -0.84732 -1.62090 -1.01450   -1.37983  0.40148
3    -0.95197 -0.14882 -0.80615 -0.01928  0.59042  0.58489   -1.37983 -1.18084
4     0.49788  0.73545 -1.63340 -0.45174 -0.30172  1.30612   -0.21712 -0.21575
...       ...      ...      ...      ...      ...      ...        ...      ...
1880 -0.95197 -1.19430  1.74091  1.88511  0.76096 -1.13788    0.88113  1.92173
1881 -0.95197 -0.24649  1.74091  0.58331  0.76096 -1.51840    0.88113      NaN
1882 -0.07854  1.13281 -1.37639 -1.27553 -1.77200 -1.38502    0.52975 -0.52593
1883 -0.95197  0.91093 -1.92173  0.29338 -1.62090 -2.57309    1.29221  1.22470
1884 -0.95197 -0.46725  2.12700  1.65653  1.11406  0.41594    0.88113  1.22470

[1885 rows x 8 columns]

time: 32.1 ms


In [0]:
df_nan.isna().sum()

Age           0
Nscore        0
Escore        0
Oscore        0
Ascore       10
Cscore       14
Impulsive     8
SS           18
dtype: int64

time: 5.27 ms


In [0]:
set(df_nan.columns)

{'Age', 'Ascore', 'Cscore', 'Escore', 'Impulsive', 'Nscore', 'Oscore', 'SS'}

time: 2.78 ms


In [0]:
missing_columns = ['Ascore', 'Cscore', 'Impulsive', 'SS']
deter_data = pd.DataFrame(columns = ["Det" + name for name in missing_columns])
from sklearn.linear_model import LinearRegression
for feature in missing_columns:
  # print(feature)
  deter_data["Det" + feature] = df_nan[feature]
  parameters = list(set(df_nan.columns) - set(missing_columns) - {feature})
  print(parameters)
  # #Create a Linear Regression model to estimate the missing data
  model = LinearRegression()
  model.fit(X = df2[parameters], y = df2[feature])

  #observe that I preserve the index of the missing data from the original dataframe
  deter_data.loc[df_nan[feature].isna(), "Det" + feature] = model.predict(df_nan[parameters])[df_nan[feature].isna()]

['Escore', 'Age', 'Oscore', 'Nscore']
['Escore', 'Age', 'Oscore', 'Nscore']
['Escore', 'Age', 'Oscore', 'Nscore']
['Escore', 'Age', 'Oscore', 'Nscore']
time: 47.7 ms


In [0]:
deter_data

DetAscore  DetCscore  DetImpulsive    DetSS
0      -0.91699   -0.00665      -0.21712 -1.18084
1       0.76096   -0.14277      -0.71126 -0.21575
2      -1.62090   -1.01450      -1.37983  0.40148
3       0.59042    0.58489      -1.37983 -1.18084
4      -0.30172    1.30612      -0.21712 -0.21575
...         ...        ...           ...      ...
1880    0.76096   -1.13788       0.88113  1.92173
1881    0.76096   -1.51840       0.88113  0.70017
1882   -1.77200   -1.38502       0.52975 -0.52593
1883   -1.62090   -2.57309       1.29221  1.22470
1884    1.11406    0.41594       0.88113  1.22470

[1885 rows x 4 columns]

time: 17.5 ms


In [0]:
non_missing_columns = ['Age', 'Nscore', 'Escore', 'Oscore', 'Ascore', 'Cscore', 'Impulsive', 'SS']
data_non_nan = pd.DataFrame(columns = ["Det" + name for name in non_missing_columns])

for feature in non_missing_columns:
  data_non_nan["Det" + feature] = df2[feature]

for feature in missing_columns:
  data_non_nan["Det" + feature] = deter_data["Det" + feature]
data_non_nan

DetAge  DetNscore  DetEscore  ...  DetCscore  DetImpulsive    DetSS
0     0.49788    0.31287   -0.57545  ...   -0.00665      -0.21712 -1.18084
1    -0.07854   -0.67825    1.93886  ...   -0.14277      -0.71126 -0.21575
2     0.49788   -0.46725    0.80523  ...   -1.01450      -1.37983  0.40148
3    -0.95197   -0.14882   -0.80615  ...    0.58489      -1.37983 -1.18084
4     0.49788    0.73545   -1.63340  ...    1.30612      -0.21712 -0.21575
...       ...        ...        ...  ...        ...           ...      ...
1880 -0.95197   -1.19430    1.74091  ...   -1.13788       0.88113  1.92173
1881 -0.95197   -0.24649    1.74091  ...   -1.51840       0.88113  0.70017
1882 -0.07854    1.13281   -1.37639  ...   -1.38502       0.52975 -0.52593
1883 -0.95197    0.91093   -1.92173  ...   -2.57309       1.29221  1.22470
1884 -0.95197   -0.46725    2.12700  ...    0.41594       0.88113  1.22470

[1885 rows x 8 columns]

time: 42 ms


In [0]:
#check error
import sklearn.metrics as metrics

mae = metrics.mean_absolute_error(X, np.array(data_non_nan))
mse = metrics.mean_squared_error(X, np.array(data_non_nan))
rmse = np.sqrt(mse) # or mse**(0.5)

print("Results of sklearn.metrics:")
print("MAE:",mae)
print("MSE:", mse)
print("RMSE:", rmse)

Results of sklearn.metrics:
MAE: 0.0017147880549502653
MSE: 0.0016718871462969177
RMSE: 0.04088871661347318
time: 7.55 ms


#Summary

##Drug Consumtion

###KNN


####Results of sklearn.metrics:
####MAE: 0.0016458772877984084
####MSE: 0.0014636197483123177
####RMSE: 0.03825728359818974

###Bayesian Network

####Results of sklearn.metrics:
####MAE: 0.002153342838196286
####MSE: 0.0026644241845424407
####RMSE: 0.05161806064298077

###Linear Regression

####Results of sklearn.metrics:
####MAE: 0.0017147880549502653
####MSE: 0.0016718871462969177
####RMSE: 0.04088871661347318